lib/core/test/unit/Network/Wai/Middleware/LoggingSpec.hs

```diff
@@ -0,0 +1,301 @@
+            let warpSettings = Warp.defaultSettings
+                    & setBeforeMainLoop (putMVar mvar p)
+            start logSettings warpSettings (traceInTVarIO tvar) socket
+        p <- readMVar mvar
+        return $ Context
+            { logs = tvar
+            , manager = mngr
+            , port = p
+            , server = handle
+            }
+
+    clearLogs :: Context -> IO ()
+    clearLogs = atomically . flip writeTVar [] . logs
+
+    tearDown :: Context -> IO ()
+    tearDown = cancel . server
+
+data Context = Context
+    { logs :: TVar [LogObject Text]
+    , manager :: Manager
+    , port :: Int
+    , server :: Async ()
+    }
+
+{-------------------------------------------------------------------------------
+                                Test Helpers
+-------------------------------------------------------------------------------}
+
+get :: Context -> String -> IO ()
+get ctx path = do
+    req <- parseRequest ("http://localhost:" <> show (port ctx) <> path)
+    void $ httpLbs req (manager ctx)
+
+delete :: Context -> String -> IO ()
+delete ctx path = do
+    req <- parseRequest ("http://localhost:" <> show (port ctx) <> path)
+    void $ httpLbs (req { method = "DELETE" }) (manager ctx)
+
+post :: ToJSON a => Context -> String -> a -> IO ()
+post ctx path json = do
+    let body = RequestBodyLBS $ Aeson.encode json
+    req <- parseRequest ("http://localhost:" <> show (port ctx) <> path)
+    void $ httpLbs (req
+        { method = "POST"
+        , requestBody = body
+        , requestHeaders = [(hContentType, "application/json")]
+        }) (manager ctx)
+
+postIlled :: Context -> String -> ByteString -> IO ()
+postIlled ctx path body = do
+    req <- parseRequest ("http://localhost:" <> show (port ctx) <> path)
+    void $ httpLbs (req
+        { method = "POST"
+        , requestBody = RequestBodyBS body
+        , requestHeaders = [(hContentType, "application/json")]
+        }) (manager ctx)
+
+expectLogs :: Context -> [(Severity, String)] -> IO ()
+expectLogs ctx expectations = do
+    entries <- reverse <$> readTVarIO (logs ctx)
+
+    when (length entries /= length expectations) $
+        fail $ "Expected exactly " <> show (length expectations)
+            <> " log entries but got " <> show (length entries) <> ": "
+            <> show entries
+
+    forM_ (zip entries expectations) $ \(l, (sev, str)) -> do
+        (severity . loMeta $ l) `shouldBe` sev
+        case loContent l of
+            LogMessage txt ->
+                T.unpack txt `shouldContain` str
+            _ ->
+                fail $ "Given log object isn't a log message but: " <> show l
+
+{-------------------------------------------------------------------------------
+                                mock server
+-------------------------------------------------------------------------------}
+
+data MkJson = MkJson
+    { field :: String
+    , sensitive :: Int
+    } deriving Generic
+
+instance ToJSON MkJson where
```

Comment:
I wonder if we have MkJson deriving Generic and we have 
```
{-# LANGUAGE DeriveGeneric #-}

import GHC.Generics
```
aren't below instances equal to
```
instance ToJSON MkJson
instance FromJSON MkJson
```
???

Reply:
Indeed ^.^... I think I just like being explicit. Force of the habit.